### PR TITLE
Fix destructor ordering for cuda handle pools

### PR DESCRIPTION
**Pull request by peterbell10**
Possible fix for gh-38385. Unfortunately, I haven't been able to reproduce the issue reliably, so can't say for certain.

Since this appears to be a destruction ordering issue, I've focused on making the destructor calls well-ordered:
- Each pool is now a function-local `static` instead of a global variable. This ensures the destructor happens before any relevant pytorch global state is destroyed.
- Each pool window now only stores a `std::weak_ptr` to the global pool. This means it can't extend the lifetime of the pool outside of the normal destructor ordering. That does also mean that if the `weak_ptr` is invalid, the handles will get leaked. However, that shouldn't happen under normal use.

### CODE REVIEW

**Comment by ezyang** [comment]
@mcarilli can you review this initially?

**Comment by mcarilli** [comment]
It's a tough problem.  I like this PR's idea's but I fear there's an antipattern somewhere else in Pytorch and we may end up leaking the handles more often than you expect.  Maybe that's ok 🤷 

Here's the order we want as the program exits:
1. Thread local `myPoolWindows` are destroyed, releasing handles back to global `DeviceThreadHandlePools<xyz>`s.
2. Global `DeviceThreadHandlePool<xyz>`s are destroyed, calling cu*Destroy on the handles.
3. Global Cuda state is torn down.

The meyers singleton for each `DeviceThreadHandlePool<xyz>` is great and I'm embarrassed I didn't do it to begin with.  Lazy-initing each global Pool ensures Pools are constructed after any global cuda state, which should mean Pools are [destroyed before](https://en.cppreference.com/w/cpp/utility/program/exit) any global cuda state.  In other words this PR's diffs help ensure 2->3 ordering.

1 and 2 [should also occur](https://stackoverflow.com/questions/43083856/are-global-objects-guaranteed-to-be-destructed-after-all-thread-local-storage-ob) in the right order if all side threads are joined before global variables start getting destroyed.

However, in https://github.com/pytorch/pytorch/pull/36416 @VivekPanyam saw segfaults that appeared to be caused by `PoolWindows` outliving their `DeviceThreadHandlePools`.  On destruction, the `PoolWindows` would try to release handles back to a `DeviceThreadHandlePool` that had already been deallocated, and segfault.

Vivek pointed out if the destructor of some global object IS the spot where some threads are joined, the threads' terminations occur somewhere amidst the destruction of global objects, and all bets are off:  the global-object destructor that joins the threads may occur after the global `DeviceThreadHandlePool<xyz>`'s destructor.  This meant the `PoolWindows` could be destroyed  after the `DeviceThreadHandlePool`.  Vivek provided minimal examples that demonstrated his concern.  

The antipattern I fear is that somewhere, some threads are being joined in the destructor of a global object, which makes it difficult to enforce 1->2 ordering.

To help ensure 1->2 ordering, Vivek's PR wrapped each global DeviceThreadHandlePool in a `shared_ptr`, and had each `thread_local` PoolWindow also store `shared_ptr`s to the global Pool, so that the `DeviceThreadHandlePool` itself would survive as long as at least one `thread_local` `shared_ptr` or the global `shared_ptr` remained.  This got rid of his segfaults.

The point I'm trying to make is, if (before adding the shared_ptrs) Vivek saw frequent segfaults caused by the thread local `PoolWindows` outliving their `DeviceThreadHandlePools`, after your PR's diffs we will leak the handles whenever that occurs.

**Comment by mcarilli** [comment]
Wait a minute, the handles aren't leaked, in the sense that `cu*Destroy` is always gracefully called on them as long as 2->3 ordering is assured.  With this PR's diffs, we don't care at all where 1 occurs with respect to 2 and 3.

There are two things each `DeviceThreadHandlePool` tracks:
```c++
    std::unordered_map<int, std::vector<Handle>> created_handles;
    std::unordered_map<int, std::vector<Handle_t>> available_handles;
```
`created_handles` tracks all the handles that have been created this run, as instances of the `Handle` wrapper that wraps a raw `cu*Handle_t`.  `Handle`'s destructor calls `cu*Destroy` on its raw `cu*Handle_t`.
`available_handles` stores lightweight `cu*Handle_t` objects (effectively pointers, probably [literally pointers](https://docs.nvidia.com/deeplearning/sdk/cudnn-api/index.html#cudnnHandle_t), intended to be passed around by value and still refer to the same underlying opaque data structure/allocation that the library calls need.

`myPoolWindow::reserve()` appends to `created_handles` if need be, but never takes a full fledged `Handle` object for itself.  It only stashes lightweight `cu*Handle_t`s during `reserve()`, then on destruction (thread exit) `release()` puts them back into `available_handles` for other threads to use.  If, at the end of a run, `DeviceThreadHandlePool` has already been destructed when `release()` is called, not putting this `PoolWindow`'s `cu*Handle_t`s back into `available_handles` is harmless.  No one's going to need them again, and `created_handles`'s destructor already took care of calling `cu*Destroy` on (a copy of) all the `cu*Handle_t`s.

tldr I think this PR is a good approach, nice work!

@VivekPanyam Before we land this, can you pull in the diffs and confirm they don't resurrect your segfaults?  I don't see why they would, but I want to be sure.

File: aten/src/ATen/cuda/CublasHandlePool.cpp [hunk 1]
```diff
@@ -21,23 +21,28 @@ void destroyCublasHandle(cublasHandle_t handle) {
...
   AT_CUDA_CHECK(cudaGetDevice(&device));
 
-  if (!myPoolWindow)
-    myPoolWindow.reset(pool->newPoolWindow());
+  // Thread local PoolWindows are wrapped by unique_ptrs and lazily-initialized
+  // to avoid initialization issues that caused hangs on Windows.
+  // See: https://github.com/pytorch/pytorch/pull/22405
+  // This thread local unique_ptrs will be destroyed when the thread terminates,
+  // releasing its reserved handles back to the pool.
+  thread_local std::unique_ptr<CuBlasPoolType::PoolWindow> myPoolWindow(
```
Comment (mcarilli) [inline aten/src/ATen/cuda/CublasHandlePool.cpp:43]: per my [first comment](https://github.com/pytorch/pytorch/pull/39345#issuecomment-643593751), I don't think moving `myPoolWindow`s from a file scope thread local to a function scope thread local actually helps enforce 1->2 ordering, but per my [second comment](https://github.com/pytorch/pytorch/pull/39345#issuecomment-643601421) in which I came to my senses, now that `PoolWindow`s hold `weak_ptr`s to their `DeviceThreadHandlePool`s, I don't think we care about 1->2 ordering at all!

So i think it's fine to construct `myPoolWindow`s in `getCurrent*Handle`s, but I also think it would be fine to leave them as file scope thread locals.  whatever you prefer.

File: aten/src/ATen/cuda/CuSparseHandlePool.cpp [hunk 2]
```diff
@@ -21,23 +21,26 @@ void destroyCusparseHandle(cusparseHandle_t handle) {
...
-auto pool = std::make_shared<DeviceThreadHandlePool<cusparseHandle_t, createCusparseHandle, destroyCusparseHandle>>();
+using CuSparsePoolType = DeviceThreadHandlePool<cusparseHandle_t, createCusparseHandle, destroyCusparseHandle>;
 
-// Thread local PoolWindows are wrapped by unique_ptrs and lazily-initialized
-// to avoid initialization issues that caused hangs on Windows.
-// See: https://github.com/pytorch/pytorch/pull/22405
-// This thread local unique_ptrs will be destroyed when the thread terminates,
-// releasing its reserved handles back to the pool.
-thread_local std::unique_ptr<decltype(pool)::element_type::PoolWindow> myPoolWindow;
+CuSparsePoolType &getPool() {
```
Comment (ezyang) [inline aten/src/ATen/cuda/CuSparseHandlePool.cpp:26]: This isn't thread local anymore

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/cuda/CuSparseHandlePool.cpp (hunk 2 above)
Comment (agolynski) [inline aten/src/ATen/cuda/CuSparseHandlePool.cpp:26]: nit: maybe give these unique names, e.g. getSparsePool, getCuBlasPool, getCuDNNPool?

**Comment by peterbell10** [comment]
I moved the `pool` static variable directly into the `get..Handle` function as it's slightly less code and avoids the function naming issues entirely.

**Review by VivekPanyam — COMMENTED** [review]
The PR looks good to me!

Doing some local testing with this change now and will comment again in a bit

**Review by VivekPanyam — APPROVED** [review]
> @VivekPanyam Before we land this, can you pull in the diffs and confirm they don't resurrect your segfaults? I don't see why they would, but I want to be sure.

Confirmed! Thanks for the heads up about this change @mcarilli
